### PR TITLE
refactor(functions): LRU eviction for admin status cache + BoundedLruMap utility

### PR DESCRIPTION
**Pull request by OPS-PIvers**
## Summary

- Switches `cachedAdminStatus` (in `functions/src/index.ts`) from FIFO to LRU eviction, so a frequently-accessed admin written early no longer gets evicted by one-off lookups.
- Extracts the pattern into a small `BoundedLruMap` utility at `functions/src/utils/boundedLruMap.ts` for the next caller (likely the student-pseudonym path, which is higher-cardinality and where FIFO would actively bite).
- Adds a regression test pinning LRU promotion: a key re-touched after the cache fills survives subsequent eviction pressure, while an untouched neighbor gets evicted instead.

Flagged by the pr-review-toolkit:code-reviewer on PR #1597 as LOW priority / acceptable-as-is at current school-district scale (~500 admins). This is the documented "chore when convenient" follow-up.

## Test plan

- [x] `pnpm exec vitest run` in `functions/` — 244/244 pass, including the new LRU promotion test and the updated eviction test.
- [x] `pnpm run validate` at the repo root — type-check (root + functions), lint (0 errors, 0 warnings), format-check, root tests (2370 pass), functions tests (244 pass).
- [ ] No runtime behavior change at current scale — admin cache hit rate is the only thing affected, and it should be marginally better. Verify in dev-preview if desired.

---
_Generated by [Claude Code](https://claude.ai/code/session_01Jyi4UKpoZYVJc2BT79igCS)_

### CODE REVIEW

File: functions/src/utils/boundedLruMap.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  get(key: K): V | undefined {
+    const value = this.map.get(key);
+    if (value === undefined) {
+      // `has` would be more correct for `V = undefined`, but the extra
+      // lookup isn't worth it for the cache values we actually store
+      // (objects, primitives that callers wrap in objects).
+      return undefined;
+    }
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply functions/src/utils/boundedLruMap.ts]: Valid — fixed in e145b86. `get()` now uses `this.map.has(key)` to distinguish present-with-`undefined` from absent, so promotion fires correctly when `V = undefined`. The previous trade-off comment is gone since the cost is one extra hash lookup per cache hit (negligible). Added a unit test (`boundedLruMap.test.ts`) that pins this with a `BoundedLruMap<string, number | undefined>` and a stored-`undefined` entry surviving eviction pressure.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: functions/src/utils/boundedLruMap.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    this.map.set(key, value);
+    return value;
+  }
+
+  set(key: K, value: V): void {
+    if (this.map.has(key)) {
+      this.map.delete(key);
+    } else if (this.map.size >= this.maxSize) {
+      const oldest = this.map.keys().next().value;
+      if (oldest !== undefined) this.map.delete(oldest);
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply functions/src/utils/boundedLruMap.ts]: Valid — fixed in e145b86. `set()` eviction now uses the iterator's `done` sentinel (`const { value: oldest, done } = this.map.keys().next(); if (!done) this.map.delete(oldest);`) instead of comparing the key to `undefined`. New unit test covers a `BoundedLruMap<string | undefined, number>` where the oldest key *is* `undefined` and confirms the map stays at `maxSize`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: functions/src/utils/boundedLruMap.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply functions/src/utils/boundedLruMap.ts:27]: Valid — fixed in e145b86. Both `get()` and `set()` now handle `V = undefined` / `K = undefined` correctly (see the replies on gemini-code-assist's two threads for the specifics). New `boundedLruMap.test.ts` pins both contracts directly, independent of the `cachedAdminStatus` caller.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by OPS-PIvers — COMMENTED** [review]
## Automated Code Review

**Date:** 2026-05-13

### Summary
Swaps `cachedAdminStatus` from FIFO to LRU eviction by extracting the pattern into a new generic `BoundedLruMap<K, V>` utility at `functions/src/utils/boundedLruMap.ts` (59 lines + 73 lines of tests). Adds a regression test on the admin-status cache that pins LRU promotion: an entry re-touched after the cap fills survives subsequent eviction pressure, while an untouched neighbor gets evicted. Tracks back to the LOW-priority `pr-review-toolkit:code-reviewer` flag on PR #1597.

### Code Quality
- ✓ `BoundedLruMap` is a clean small utility — constructor validates `maxSize` is a positive integer, `get` uses `this.map.has(key)` (so an explicit `undefined` value is a hit), `set` uses the iterator `done` sentinel (so an `undefined` key still gets evicted correctly). Both edge cases were flagged by reviewers and addressed in e145b86 with dedicated tests.
- ✓ Re-insert-after-delete is the standard idiom for Map-backed LRU; no `console.log`, no `any`, no floating promises.
- ✓ Eviction loop replaced with a single call to `this.map.set` — the old `if (cachedAdminStatus.size >= ADMIN_STATUS_CACHE_MAX)` block in `index.ts` (lines 263–274 of the pre-PR file) is gone, simplifying the caller.

### Widget Standards
✓ Not applicable — Cloud Functions only.

### Admin Config Standards
✓ Not applicable.

### Firestore & Security
- ✓ Cache key is `emailLower` (already lowercased), value is `{ isAdmin, cachedAt }`. No PII change, no Firestore rule change. Cache TTL semantics unchanged from before — the LRU swap only changes *which* entries get evicted, not *whether* they expire.
- ✓ `cachedAdminStatus.size` direct reads from the old code are gone — `BoundedLruMap` exposes `size` as a getter, which the new code uses for boundary assertions in tests.

### Regression Risk
- **Low.** Single-file abstraction; the caller-side change in `functions/src/index.ts` is +8 / -12 with no other behavioral coupling.
- ✓ The new `boundedLruMap.test.ts` covers six scenarios: `maxSize` validation, missing-key miss, promote-on-read survival, explicit-`undefined`-value hit + promotion, `undefined`-key eviction, and `delete`/`clear` semantics.
- ✓ The existing `'evicts the oldest entry when the admin cache exceeds its bound'` test was updated to match LRU semantics (renamed to `'evicts the least-recently-used entry…'`), and the new `'LRU: a recently-accessed key survives…'` test specifically asserts the FIFO-vs-LRU divergence (510 vs 511 reads after promotion).
- ✓ `pnpm exec vitest run` in `functions/` reported 244/244 passing per the PR description; full `pnpm run validate` clean at the root.

### Missing Tests
✓ Coverage is solid for both `BoundedLruMap` (6 unit tests) and the integration (the existing-renamed test + the new LRU-survival test in `functions/src/index.test.ts`).

### Overall
**Merge readiness:** Ready
Textbook small-refactor PR: replace one pattern with a slightly better one, extract for reuse, lock in the new contract with focused unit tests, and address every reviewer edge case (the `K = undefined` / `V = undefined` paths in particular). The reusability framing — "next caller is likely the higher-cardinality student-pseudonym path" — is the right justification for promoting this from inline code to a named utility.

---
_This review was generated automatically. All findings should be verified by a human before merging._

---
_Generated by [Claude Code](https://claude.ai/code)_
